### PR TITLE
feat(metrics): add server google analytics events

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes https://github.com/mozilla/fxa-content-server/issues/2782

### CODE REVIEW

**Comment by vladikoff** [comment]
@shane-tomlinson @zaach r?

File: server/lib/ga-collector.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+   * Send a formatted metrics object to Google Analytics
+   *
+   * @param {Object} body
+   */
+  write: function (body) {
+    if (! body || ! body.events || ! body.events.length > 0 || ! this.analyticsId) {
+      return;
+    }
+
+    var visitor = ua(ANALYTICS_ID, { https: true, debug: true });
```
Comment (shane-tomlinson) [inline server/lib/ga-collector.js]: I read `ua` here as `user-agent` and was confused. Could you change this to a more descriptive name?
Do you want `debug: true`?

File: server/lib/ga-collector.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    body.events.forEach(function (event) {
+      if (event.type && Object.prototype.hasOwnProperty.call(GA_EVENTS, event.type)) {
```
Comment (shane-tomlinson) [inline server/lib/ga-collector.js]: We know `GA_EVENTS` exists, it's defined in this file & the extra `Object.prototype.hasOwnProperty` doesn't really seem necessary. Could you just do a simple `GA_EVENTS[event.type]`?

File: server/lib/ga-collector.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        var gaEvent = GA_EVENTS[event.type];
+
+        // see https://github.com/peaksandpies/universal-analytics/blob/master/AcceptableParams.md
+        // for available list of parameters
+        var gaData = {
+          hitType: 'event',
```
Comment (shane-tomlinson) [inline server/lib/ga-collector.js]: Is it super pedantic to ask for this list to be alphabetized? Yes, but still. ;)

File: server/lib/routes/post-metrics.js [hunk 4]
```diff
@@ -26,6 +28,8 @@ module.exports = function () {
         // send the metrics body to the StatsD collector for processing
         statsd.write(metrics);
       }
+
```
Comment (shane-tomlinson) [inline server/lib/routes/post-metrics.js]: Do you think we should `process.nextTick` everything after `res.json`?

File: tests/server/metrics-ga.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          return this;
+        },
+        send: function () {}
+      };
+    }
+
+    var mocks = {
+      'universal-analytics': analyticsMock
+    };
+    var GACollector = proxyquire(process.cwd() + '/server/lib/ga-collector', mocks);
```
Comment (shane-tomlinson) [inline tests/server/metrics-ga.js]: Is the convention to use `path.join` to create a path?

File: tests/server/metrics-ga.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+          assert.equal(data.ev, 1);
+          assert.equal(data.hitType, 'event');
+          assert.equal(data.anonymizeIp, 1);
+          assert.equal(data.dataSource, 'web');
+          assert.isDefined(data.ua);
+          assert.equal(data.campaignName, 'none');
+          assert.equal(data.campaignSource, 'none');
+          assert.equal(data.campaignMedium, 'none');
+          assert.isDefined(data.documentHostName, 'none');
+          assert.equal(data.uid, 'c614d7fb-43e4-485c-bf11-40afbb202656');
```
Comment (shane-tomlinson) [inline tests/server/metrics-ga.js]: Do you expect the `cid` and `uid` to be the same?

File: tests/server/metrics-ga.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    function analyticsMock() {
+      return {
+        event: function (data) {
+          assert.equal(data.ec, SIGNUP_FLOW);
+          assert.equal(data.ea, 'registered');
+          assert.equal(data.el, 'regular');
+          assert.equal(data.ev, 1);
+          assert.equal(data.hitType, 'event');
+          assert.equal(data.anonymizeIp, 1);
```
Comment (shane-tomlinson) [inline tests/server/metrics-ga.js]: Might be nice to alphabetize these too.

File: server/lib/ga-collector.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+var config = require('./configuration');
+
+var PUBLIC_URL = config.get('public_url');
+var ANALYTICS_ID = config.get('google_analytics_id');
+var SIGNUP_FLOW = 'Firefox Accounts Sign-up Flow';
+
+// note: if you use labels (el), then a value must be present (ev).
+var GA_EVENTS = {
+  'screen.signup': {
+    ec: SIGNUP_FLOW,
```
Comment (shane-tomlinson) [inline server/lib/ga-collector.js]: Is there some documentation on what ec/ea/el/ev mean & why some are optional?

File: server/lib/ga-collector.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+GACollector.prototype = {
+  /**
+   * Send a formatted metrics object to Google Analytics
+   *
+   * @param {Object} body
+   */
+  write: function (body) {
+    if (! body || ! body.events || ! body.events.length > 0 || ! this.analyticsId) {
```
Comment (shane-tomlinson) [inline server/lib/ga-collector.js]: Removing the 'google_analytics_id' from config is an easy kill switch if we need to disable GA reporting. Good.

File: app/scripts/lib/app-start.js [hunk 10]
```diff
@@ -275,6 +275,7 @@ function (
         screenWidth: screenInfo.screenWidth,
         able: this._able,
         isSampledUser: isSampledUser,
+        uniqueUserId: this._getUniqueUserId(),
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: Not for this PR, but we are now passing in both `able` and the `uniqueUserId`. We could move the [isSampledUser](https://github.com/mozilla/fxa-content-server/pull/2787/files#diff-b80f22cfd726dfdc2bb7ed3e82dc7264R257) logic into the Metrics module.

Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: If we did that, we'd also have to pass `env`. Nevermind.

File: tests/server/metrics-ga.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+          assert.equal(data.dataSource, 'web');
+          assert.isDefined(data.ua);
+          assert.equal(data.campaignName, 'none');
+          assert.equal(data.campaignSource, 'none');
+          assert.equal(data.campaignMedium, 'none');
+          assert.isDefined(data.documentHostName, 'none');
```
Comment (philbooth) [inline tests/server/metrics-ga.js]: This `'none'` looks rogue.

File: server/lib/routes/post-metrics.js (hunk 4 above)
Comment (vladikoff) [inline server/lib/routes/post-metrics.js]: sure!

File: server/lib/ga-collector.js (hunk 8 above)
Comment (vladikoff) [inline server/lib/ga-collector.js]: Will add some links

File: tests/server/metrics-ga.js (hunk 6 above)
Comment (vladikoff) [inline tests/server/metrics-ga.js]: yeah, seems like the difference is: client id === identify "session, user id === logged in user. we don't want to save the logged in user so we can use the uuid for both. See http://cutroni.com/blog/2014/04/10/understanding-cross-device-measurement-and-the-user-id/ and https://github.com/peaksandpies/universal-analytics/blob/master/AcceptableParams.md#client-id

**Comment by vladikoff** [comment]
@shane-tomlinson @zaach r?

**Comment by shane-tomlinson** [comment]
ref #2101

**Comment by shane-tomlinson** [comment]
Here is what I see in GA when running the functional tests:

screen shot 2015-07-21 at 17 04 23

**Pull request merged**